### DISCUSSION TITLE
좋아하는 자료구조

### DISCUSSION THREAD

**Discussion by WooGie911** [wanted-9th-3team/frontend-study#52] [프론트 기술 질문 공간] [answered]
좋아하는 자료구조가 있다면 이유와 함께 설명해주세요.
없으시다면 아무 자료구조에 대해 설명해주세요.

**Comment by michoball** [comment]
linked list 

데이터를 가진 노드와 다음 리스트의 위치 정보를 가지고 있는 링크로 이뤄진 자료구조 입니다. 
새로운 노드를 추가하거나 삭제하는데 배열보다 효율적으로 작동합니다. 
( 추가 - 맨 뒤에  null 을 가르키는 링크를 가진 노드에 새 노드를 가르키도록 연결)
( 삭제 - 해당 노드의 앞 노드의 링크를 해당 노드 뒤 노드를 가르키도록 하면 삭제 )

동적으로 크기조절이 가능하기에 메모리를 효율적으로 사용하는 자료구조입니다.

싱글 링크드 리스트의 경우는 다음 노드의 정보만을 가지기 때문에 역방향으로의 순회에는 불리합니다. 
그런경우 더블 링크드 리스트로 이전 노드의 정보를 가진 추가 링크를 가진 링크드 리스트를 사용할 수 있지만 싱글에 비해 메모리의 사용량이 많습니다. 

이전 노드의 정보와 본인의 정보만을 가지고 있다 특징 때문에 블록체인에 사용되는 자료구조로 알고 있습니다.

**Comment by WooGie911** [comment, marked as answer]
저는 개인적으로 linked list를 좋아합니다. 
왜냐하면 배열에 비해 삽입, 삭제가 용이하고, 트리 구조의 근간이 된다고 볼 수 있기 때문입니다.

linked list에 대해 간단하게 설명하면 linked list는 여러 개의 노드들이 순차적으로 연결된 형태를 갖는 자료구조이며, 첫번째 노드를 head, 마지막 노드를 tail이라고 합니다. 
각 노드는 데이터와 다음 노드를 가리키는 포인터로 이루어져 있으며 배열과는 다르게 메모리를 연속적으로 사용하지 않습니다. 

배열과 비교 했을 때 원하는 위치에 한번에 접근할 수 없어 순차적으로 접근해야 하는 면에서 불리할 수도 있으나, 노드가 연결된 구조이기 때문에 삽입, 삭제에 용이합니다. (배열은 입력된 데이터들이 메모리 공간에서 연속적으로 저장되어 있는 자료구조로 index를 통한 접근이 용이하며 배열의 크기는 처음 생성할 때 정하며 이후에 변경할 수 없습니다. )

또한 Tree 구조의 근간이 되는 자료구조이며, Tree 에서 사용되었을 때 그 유용성이 드러납니다.

**Comment by BaekYuri** [comment]
중복 값을 처리할 때, set을 즐겨 사용하는 편입니다.
Set은 순서가 없고, 중복된 값을 저장하지 않습니다. 그렇기 때문에 어떤 배열을 순회하면서 중복된 값만 저장할 때나, 어떠한 값을 검색할 때 빠른 속도로 찾을 수 있어 좋습니다.
